// eas-cli/build/worker/utils/multipart.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createMultipartBodyFromFilesAsync = exports.multipartContentType = void 0;
const tslib_1 = require("tslib");
const node_fs_1 = tslib_1.__importDefault(require("node:fs"));
const CRLF = '\r\n';
const BOUNDARY_HYPHEN_CHARS = '--';
const BOUNDARY_ID = '----formdata-eas-cli';
const FORM_FOOTER = `${BOUNDARY_HYPHEN_CHARS}${BOUNDARY_ID}${BOUNDARY_HYPHEN_CHARS}${CRLF}${CRLF}`;
const encodeName = (input) => {
    return input.replace(/["\n\\]/g, (c) => {
        switch (c) {
            case '\\':
                return '\\\\';
            case '"':
                return '%22';
            case '\n':
                return '%0A';
            default:
                return `%${c.charCodeAt(0).toString(16).toUpperCase()}`;
        }
    });
};
async function* createReadStreamAsync(filePath) {
    for await (const raw of node_fs_1.default.createReadStream(filePath)) {
        const chunk = raw;
        yield new Uint8Array(chunk.buffer, chunk.byteOffset, chunk.byteLength);
    }
}
const makeFormHeader = (params) => {
    const name = encodeName(params.name);
    let header = BOUNDARY_HYPHEN_CHARS + BOUNDARY_ID + CRLF;
    header += `Content-Disposition: form-data; name="${name}"; filename="${name}"`;
    if (params.contentType) {
        header += `${CRLF}Content-Type: ${params.contentType}`;
    }
    if (params.contentLength) {
        header += `${CRLF}Content-Length: ${params.contentLength}`;
    }
    header += CRLF;
    header += CRLF;
    return header;
};
exports.multipartContentType = `multipart/form-data; boundary=${BOUNDARY_ID}`;
async function* createMultipartBodyFromFilesAsync(entries, onProgressUpdate) {
    const encoder = new TextEncoder();
    for (let idx = 0; idx < entries.length; idx++) {
        const entry = entries[idx];
        const header = makeFormHeader({
            name: entry.name,
            contentType: entry.contentType,
            contentLength: entry.contentLength,
        });
        yield encoder.encode(header);
        yield* createReadStreamAsync(entry.filePath);
        yield encoder.encode(CRLF);
        if (onProgressUpdate) {
            onProgressUpdate((idx + 1) / entries.length);
        }
    }
    yield encoder.encode(FORM_FOOTER);
}
exports.createMultipartBodyFromFilesAsync = createMultipartBodyFromFilesAsync;
